pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -657,216 +567,88 @@ public long getCommitTimeoutMS(String tableName) {
-        segmentStartOffset = prevSegEndOffset;
-        LOGGER.info("Choosing newer offset {} for table {} for partition {}, sequence {}", oldestOffsetInStream,
-            realtimeTableName, partition, nextSeqNum);
-      } else if (oldestOffsetInStream > prevSegEndOffset) {
-        // Stream's oldest offset is greater than the end offset of the prev segment, so there is data loss.
-        LOGGER.warn("Data lost from offset {} to {} for table {} partition {} sequence {}", prevSegEndOffset,
-            oldestOffsetInStream, realtimeTableName, partition, nextSeqNum);
-        _controllerMetrics.addMeteredTableValue(realtimeTableName, ControllerMeter.LLC_KAFKA_DATA_LOSS, 1);
-        _controllerMetrics.addMeteredTableValue(realtimeTableName, ControllerMeter.LLC_STREAM_DATA_LOSS, 1);
-      } else {
-        // The two happen to be equal. A rarity, so log it.
-        LOGGER.info("Earliest offset {} is the same as new segment start offset", oldestOffsetInStream);
-      }
-    }
-    return segmentStartOffset;
-  }
-
-  private ZNRecord makeZnRecordForNewSegment(String realtimeTableName, int numReplicas, long startOffset,
-      LLCSegmentName newSegmentName, int numPartitions) {
-    final LLCRealtimeSegmentZKMetadata newSegMetadata = new LLCRealtimeSegmentZKMetadata();
-    newSegMetadata.setCreationTime(System.currentTimeMillis());
-    newSegMetadata.setStartOffset(startOffset);
-    newSegMetadata.setEndOffset(END_OFFSET_FOR_CONSUMING_SEGMENTS);
-    newSegMetadata.setNumReplicas(numReplicas);
-    newSegMetadata.setTableName(realtimeTableName);
-    newSegMetadata.setSegmentName(newSegmentName.getSegmentName());
-    newSegMetadata.setStatus(CommonConstants.Segment.Realtime.Status.IN_PROGRESS);
-
-    // Add the partition metadata if available.
-    SegmentPartitionMetadata partitionMetadata =
-        getPartitionMetadataFromTableConfig(realtimeTableName, numPartitions, newSegmentName.getPartitionId());
-    if (partitionMetadata != null) {
-      newSegMetadata.setPartitionMetadata(partitionMetadata);
-    }
-
-    return newSegMetadata.toZNRecord();
-  }
-
   /**
    * An instance is reporting that it has stopped consuming a topic due to some error.
    * Mark the state of the segment to be OFFLINE in idealstate.
    * When all replicas of this segment are marked offline, the {@link org.apache.pinot.controller.validation.RealtimeSegmentValidationManager},
    * in its next run, will auto-create a new segment with the appropriate offset.
    */
-  public void segmentStoppedConsuming(final LLCSegmentName segmentName, final String instance) {
-    String rawTableName = segmentName.getTableName();
-    String realtimeTableName = TableNameBuilder.REALTIME.tableNameWithType(rawTableName);
-    final String segmentNameStr = segmentName.getSegmentName();
+  public void segmentStoppedConsuming(LLCSegmentName llcSegmentName, String instanceName) {
+    Preconditions.checkState(!_isStopping, "Segment manager is stopping");
```

Comment:
if the segment manger is stopping, we should return NOT_LEADER, so that pinot-server can connect to another leader and do the operation again. Same for other public operations initiated by the server.

Reply:
Discussed offline. The leader check are performed before getting into this class.